src/java.base/unix/native/libnio/fs/UnixCopyFile.c

```diff
@@ -47,15 +51,35 @@ static void throwUnixException(JNIEnv* env, int errnum) {
 }
 
 /**
- * Transfer all bytes from src to dst via user-space buffers
+ * Transfer all bytes from src to dst within the kernel if possible (Linux),
+ * otherwise via user-space buffers
  */
 JNIEXPORT void JNICALL
 Java_sun_nio_fs_UnixCopyFile_transfer
     (JNIEnv* env, jclass this, jint dst, jint src, jlong cancelAddress)
 {
-    char buf[8192];
     volatile jint* cancel = (jint*)jlong_to_ptr(cancelAddress);
 
+#if defined(__linux__)
+    // Transfer within the kernel
+    const size_t count = 1048576; // 1 MB to give cancellation a chance
+    ssize_t bytes_sent;
+    do {
+        // sendfile() can transfer at most 0x7ffff000 bytes
+        RESTARTABLE(sendfile64(dst, src, NULL, count), bytes_sent);
+        if (bytes_sent == -1) {
+            throwUnixException(env, errno);
+            return;
+        }
+        if (cancel != NULL && *cancel != 0) {
+            throwUnixException(env, ECANCELED);
+            return;
+        }
+    } while (bytes_sent > 0);
+#else
```

Comment:
Thanks, I think this right. I'll double it on Friday.

Reply:
All right, thanks.